File: StriverA2Z/C++/6_LinkedList/Questions/DoublyLinkedList/remove_duplicates.cpp
#include<bits/stdc++.h>
#include "Node.h"

using namespace std;

Node *removeDuplicates(Node *head) {
    Node *temp = head;
    while (temp && temp->next) {
        Node *nextNode = temp->next;
        while (nextNode && nextNode->data == temp->data) {
            Node *duplicate = nextNode;
            nextNode = nextNode->next;
            free(duplicate);
        }
        temp->next = nextNode;
        if (nextNode) nextNode->prev = temp;
        temp = temp->next;
    }
    return head;
}

int main() {
    Node *head = arrayToLinkedList({1,1,1,2,3,3,4});
    head = removeDuplicates(head);
    traverseToLinkedList(head);
    return 0;
}

File: StriverA2Z/Java/LinkedLists/Questions/DoublyLinkedList/RemoveDuplicates.java
public class RemoveDuplicates {
    public static void main(String[] args) {
        Node head = LinkedListOperations.arrayToLinkedList(new int[]{1, 1, 1, 2, 3, 3, 4});
        head = removeDuplicates(head);
        LinkedListOperations.traverseLinkedList(head);
    }

    static Node removeDuplicates(Node head) {
        Node temp = head;
        while (temp != null && temp.next != null) {
            Node nextNode = temp.next;
            while (nextNode != null && nextNode.data == temp.data) nextNode = nextNode.next;
            temp.next = nextNode;
            if (nextNode != null) nextNode.prev = temp;
            temp = temp.next;
        }
        return head;
    }
}
